File: boards/esp32-s3-devkitc1/src/ws2812.cpp
#include "ws2812.h"

static Adafruit_NeoPixel* ws2812Strip = nullptr;
static int ledCount = 0;

void initWS2812(int pin, int count) {
  if (ws2812Strip != nullptr) {
    delete ws2812Strip;
  }
  ledCount = count;
  ws2812Strip = new Adafruit_NeoPixel(count, pin, NEO_GRB + NEO_KHZ800);
  ws2812Strip->begin();
  ws2812Strip->show();
}

void setWS2812Pixel(uint16_t index, uint8_t red, uint8_t green, uint8_t blue) {
  if (ws2812Strip != nullptr && index < ledCount) {
    ws2812Strip->setPixelColor(index, ws2812Strip->Color(red, green, blue));
  }
}

void fillWS2812(uint8_t red, uint8_t green, uint8_t blue) {
  if (ws2812Strip != nullptr) {
    ws2812Strip->fill(ws2812Strip->Color(red, green, blue));
  }
}

void showWS2812() {
  if (ws2812Strip != nullptr) {
    ws2812Strip->show();
  }
}

void turnOffWS2812() {
  if (ws2812Strip != nullptr) {
    ws2812Strip->clear();
    ws2812Strip->show();
  }
}

void testWS2812() {
  if (ws2812Strip == nullptr) return;
  
  for (int j = 0; j < 3; j++) {
    for (int i = 0; i < ledCount; i++) {
      ws2812Strip->setPixelColor(i, ws2812Strip->Color(255, 0, 0));
    }
    ws2812Strip->show();
    delay(500);
    
    for (int i = 0; i < ledCount; i++) {
      ws2812Strip->setPixelColor(i, ws2812Strip->Color(0, 255, 0));
    }
    ws2812Strip->show();
    delay(500);
    
    for (int i = 0; i < ledCount; i++) {
      ws2812Strip->setPixelColor(i, ws2812Strip->Color(0, 0, 255));
    }
    ws2812Strip->show();
    delay(500);
  }
  
  turnOffWS2812();
}

Adafruit_NeoPixel* getWS2812() {
  return ws2812Strip;
}

File: blocklyduino/generators/arduino/playground_brumbrum/ws2812.js
/**
 * Visual Blocks Language
 *
 * Copyright 2024
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @fileoverview Generating Arduino for WS2812 LED Strip blocks.
 */
'use strict';

Blockly.Arduino.ws2812_init = function() {
  var dropdown_pin = this.getFieldValue('PIN');
  var count = Blockly.Arduino.valueToCode(this, 'COUNT', Blockly.Arduino.ORDER_ATOMIC);
  Blockly.Arduino.definitions_['include_ws2812_h'] = '#include "ws2812.h"\n';
  var code = 'initWS2812(' + dropdown_pin + ', ' + count + ');  // Initialize WS2812 LED strip with ' + count + ' LEDs on pin ' + dropdown_pin + '\n';
  return code;
};

Blockly.Arduino.ws2812_set_pixel = function() {
  Blockly.Arduino.definitions_['include_ws2812_h'] = '#include "ws2812.h"\n';
  var index = Blockly.Arduino.valueToCode(this, 'INDEX', Blockly.Arduino.ORDER_ATOMIC);
  var red = Blockly.Arduino.valueToCode(this, 'RED', Blockly.Arduino.ORDER_ATOMIC);
  var green = Blockly.Arduino.valueToCode(this, 'GREEN', Blockly.Arduino.ORDER_ATOMIC);
  var blue = Blockly.Arduino.valueToCode(this, 'BLUE', Blockly.Arduino.ORDER_ATOMIC);
  var code = 'setWS2812Pixel(' + index + ', ' + red + ', ' + green + ', ' + blue + ');\n';
  return code;
};

Blockly.Arduino.ws2812_fill = function() {
  Blockly.Arduino.definitions_['include_ws2812_h'] = '#include "ws2812.h"\n';
  var red = Blockly.Arduino.valueToCode(this, 'RED', Blockly.Arduino.ORDER_ATOMIC);
  var green = Blockly.Arduino.valueToCode(this, 'GREEN', Blockly.Arduino.ORDER_ATOMIC);
  var blue = Blockly.Arduino.valueToCode(this, 'BLUE', Blockly.Arduino.ORDER_ATOMIC);
  var code = 'fillWS2812(' + red + ', ' + green + ', ' + blue + ');\n';
  return code;
};

Blockly.Arduino.ws2812_show = function() {
  Blockly.Arduino.definitions_['include_ws2812_h'] = '#include "ws2812.h"\n';
  var code = 'showWS2812();\n';
  return code;
};

Blockly.Arduino.ws2812_off = function() {
  Blockly.Arduino.definitions_['include_ws2812_h'] = '#include "ws2812.h"\n';
  var code = 'turnOffWS2812();\n';
  return code;
};

Blockly.Arduino.ws2812_test = function() {
  Blockly.Arduino.definitions_['include_ws2812_h'] = '#include "ws2812.h"\n';
  var code = 'testWS2812();\n';
  return code;
};
